model-optimizer/extensions/ops/select.py

```diff
@@ -39,27 +39,46 @@ def infer(node: Node):
 
         a_shape = node.in_port(1).data.get_shape()
         b_shape = node.in_port(2).data.get_shape()
-        output_shape = bi_directional_shape_broadcasting(a_shape, b_shape)
+        broadcast_rule = node.soft_get('auto_broadcast', 'numpy')
+        if broadcast_rule == 'numpy':
+            output_shape = bi_directional_shape_broadcasting(a_shape, b_shape)
+        elif broadcast_rule == 'pdpd':
+            # todo: add pdpd broadcasting rule
+            # note that additionally to output_shape resulting_tensors must be broadcasted as well
+            raise Error("PDPD broadcasting rule is not implemented yet")
+        else:  # broadcasting is not allowed
+            assert np.array_equal(a_shape, b_shape) and condition_value is not None and np.array_equal(condition_value.shape, a_shape), \
+                'In node \'{}\' for Select operation when broadcasting is off all inputs must be of the same shape. ' \
+                'But instead got: cond_shape={}, then_shape={}, else_shape={}'.format(
+                    node_name, condition_value.shape, a_shape, b_shape)
+            output_shape = a_shape
+
         assert output_shape is not None, 'Input shapes for node {} are not broadcast-able'.format(node_name)
         node.out_port(0).data.set_shape(output_shape)
 
         if condition_value is not None:
-            if resulting_tensors[0] is not None:
-                resulting_tensors[0] = bi_directional_broadcasting(resulting_tensors[0], b_shape)
-            if resulting_tensors[1] is not None:
-                resulting_tensors[1] = bi_directional_broadcasting(resulting_tensors[1], a_shape)
-            condition_value = bi_directional_broadcasting(condition_value, output_shape)
+            if resulting_tensors[0] is None or resulting_tensors[1] is None:
+                assert np.all(condition_value == condition_value.item(0))
+                # in some graphs Select condition is always True[False] and
+                # one of the branches is None (which is not selected)
+                # if we use np.where for such cases then dtype of output_value will be object (non numeric type)
+                # and subsequent numpy operation on such tensors will fail
+                output_value = resulting_tensors[not np.bool(condition_value.item(0))]
```

Comment:
what if `condition_value` is a masked array and it equals `np.ma.array([True], mask=[1])` then `np.bool(condition_value.item(0))` is True. But output_value must be with undetermined values.

Reply:
Thanks for the valuable comment! Corrected infer function so that it correctly handles masked conditions as well. Added such cases into unit-tests